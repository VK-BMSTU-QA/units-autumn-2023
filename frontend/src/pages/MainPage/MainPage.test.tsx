import '@testing-library/jest-dom';

import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import * as useCurrentTimeHookModule from '../../hooks/useCurrentTime';
import * as applyCategoriesUtilsModule from '../../utils/applyCategories';
import * as updateCategoriesUtilsModule from '../../utils/updateCategories';
import { MainPage } from './MainPage';

const mockUseCurrentTime = jest.spyOn(
    useCurrentTimeHookModule,
    'useCurrentTime'
);
mockUseCurrentTime.mockReturnValue('12:00.00');

const mockApplyCategories = jest.spyOn(
    applyCategoriesUtilsModule,
    'applyCategories'
);
const mockUpdateCategories = jest.spyOn(
    updateCategoriesUtilsModule,
    'updateCategories'
);

afterEach(jest.clearAllMocks);

describe('MainPage test', () => {
    it('should render correctly', () => {
        const rendered = render(<MainPage />);

        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('should call useCurrentTime once on create', () => {
        render(<MainPage />);

        expect(mockUseCurrentTime).toHaveBeenCalledTimes(1);
    });

    it('should update selected categories on category click', () => {
        const rendered = render(<MainPage />);

        expect(mockUpdateCategories).toHaveBeenCalledTimes(0);

        fireEvent.click(rendered.getAllByText('Одежда')[0]);

        expect(mockUpdateCategories).toHaveBeenCalledTimes(1);

        fireEvent.click(rendered.getAllByText('Одежда')[0]);

        expect(mockUpdateCategories).toHaveBeenCalledTimes(2);
    });

    it('should apply categories to products on category click', () => {
        const rendered = render(<MainPage />);

        expect(mockApplyCategories).toHaveBeenCalledTimes(1);
        expect(
            rendered.container.getElementsByClassName(
                'categories__badge_selected'
            )
        ).toHaveLength(0);

        fireEvent.click(rendered.getAllByText('Электроника')[0]);
        expect(mockApplyCategories).toHaveBeenCalledTimes(2);
        expect(
            rendered.container.getElementsByClassName('product-card')
        ).toHaveLength(2);
        expect(
            rendered.container.getElementsByClassName(
                'categories__badge_selected'
            )
        ).toHaveLength(1);
        expect(rendered.getByText('IPhone 14 Pro')).toBeInTheDocument();
        expect(rendered.getByText('Принтер')).toBeInTheDocument();
        expect(rendered.queryByText('Костюм гуся')).not.toBeInTheDocument();

        fireEvent.click(rendered.getAllByText('Электроника')[0]);
        expect(mockApplyCategories).toHaveBeenCalledTimes(3);
        expect(
            rendered.container.getElementsByClassName(
                'categories__badge_selected'
            )
        ).toHaveLength(0);
    });
});
