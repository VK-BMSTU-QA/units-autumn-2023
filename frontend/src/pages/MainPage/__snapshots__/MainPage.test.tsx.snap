// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MainPage should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3>
      03:00:00
    </h3>
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          MacBook
        </h2>
        <div
          class="product-card__description"
        >
          M2
        </div>
        <h1
          class="product-card__price"
        >
          1 400 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
      <img
        alt="MacBook"
        class="product-card__image"
        src="url"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Bershka
        </h2>
        <div
          class="product-card__description"
        >
          T-short
        </div>
        <h1
          class="product-card__price"
        >
          14 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
      <img
        alt="Bershka"
        class="product-card__image"
        src="url"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          H&M
        </h2>
        <div
          class="product-card__description"
        >
          Trousers
        </div>
        <h1
          class="product-card__price"
        >
          15 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
      <img
        alt="H&M"
        class="product-card__image"
        src="url"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Bershka
        </h2>
        <div
          class="product-card__description"
        >
          Hat
        </div>
        <h1
          class="product-card__price"
        >
          10 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
      <img
        alt="Bershka"
        class="product-card__image"
        src="url"
      />
    </div>
  </div>
</DocumentFragment>
`;
